### ISSUE TITLE
Object expressions should allow for new interface implementations, and fit into parameters of said interface

### ISSUE THREAD

**Issue by Happypig375** [fsharp/fslang-suggestions#944]
# Object expressions should allow for new interface implementations, and fit into parameters of said interface

I propose we allow
```fs
let videoView = new Android.Views.VideoView(context)
videoView.SetOnPreparedListener({
    new Java.Lang.Object() with
        interface Android.Media.MediaPlayer.IOnPreparedListener with
            member _.OnPrepared player = () })
// Currently fails with
// FS0010 Incomplete structured construct at or before this point in expression. Expected '}' or other token.	
// FS0583 Unmatched '('	AndroidTVPushNotifications	
// FS0193 Type constraint mismatch. The type 'Java.Lang.Object' is not compatible with type 'Android.Media.MediaPlayer.IOnPreparedListener'
// FS0604 Unmatched '{'
// FS0010 Unexpected keyword 'interface' in object expression. Expected 'member', 'override' or other token.
// FS0010 Unexpected start of structured construct in object expression
```

The existing way of approaching this problem in F# is writing a boilerplate class.

## Pros and Cons 

The advantages of making this adjustment to F# are
1. Better interop with Xamarin.Android
2. Expand the use cases of object expressions
3. Help write more concise code.

The disadvantage of making this adjustment to F# is that object expressions may be too niche to be improved. However, it is a niche exactly because of cases like these where it cannot be used.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S

Related suggestions: (put links to related suggestions here)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by FunctionalFirst** [comment]
I think this may be related to the fact that it isn't possible to implement an abstract class using object expressions without overriding at least one member (see [related SO question](https://stackoverflow.com/questions/8154730/object-expression-for-abstract-class-without-abstract-members)). You can work around this limitation by overriding one of `System.Object`'s virtual members.

````f#
let videoView = new Android.Views.VideoView(context)
videoView.SetOnPreparedListener({
    new Java.Lang.Object() with
        member _.ToString() = base.ToString()
    interface Android.Media.MediaPlayer.IOnPreparedListener with
        member _.OnPrepared player = () })
````

**Comment by Happypig375** [comment]
@FunctionalFirst 
![image](https://user-images.githubusercontent.com/19922066/103775582-33749a80-5069-11eb-8387-a57dbc0b1acb.png)
(Sorry for wrong namespace btw)

**Comment by FunctionalFirst** [comment]
@Happypig375 You're right—my mistake. There were no squiggles, but I see the same error if I send to interactive.

**Comment by charlesroddie** [comment]
> The existing way of approaching this problem in F# is writing a boilerplate class.
I.e.
```fsharp
type private ListeningJavaObject() =
    inherit Java.Lang.Object()
    interface Android.Media.MediaPlayer.IOnPreparedListener with
        member _.OnPrepared player = ()
let videoView = new Android.Widget.VideoView(context)
videoView.SetOnPreparedListener(ListeningJavaObject())
```
If the intention of allowing the proposed code over the explicit class is to save one line, that not a real advantage. If the intention is to allow a more local definition, then https://github.com/fsharp/fslang-suggestions/issues/150 would be better way to do it, although a much larger feature to implement. But since there are several new issues about anonymous classes and interfaces it might be considered instead of implementing all of those.

**Comment by debugnik** [comment]
@Happypig375 The underlying object still implements the interface, so you could `unbox` the object expression, but at the expense of static type checking.

The missing feature here is object expressions having an anonymous type when they inherit/implement more than one type. As for the unnecessary override for the parent type, #632 is approved, but missing an RFC.
